File: complicaus_2/no_prefix_set.py
# Source: HackerRank
# URL: https://www.hackerrank.com/challenges/no-prefix-set/problem
# Solution by Pablo Reina Jimenez

def insert_word(trie, word):
    current = trie
    for char in word:
        if char not in current:
            current[char] = {}
        current = current[char]
        if "END" in current:
            return False

    current["END"] = True

    if len(current) > 1:
        return False

    return True


def check_good_set(words):
    trie = {}
    for word in words:
        if not insert_word(trie, word):
            print("BAD SET")
            print(word)
            return

    print("GOOD SET")


n = int(input())
words = [input().strip() for _ in range(n)]
check_good_set(words)

File: complicaus_2/no_prefix_set.hs
-- Source: HackerRank
-- URL: https://www.hackerrank.com/challenges/no-prefix-set/problem
-- Solution by Alejandro Domínguez Muñoz

import Control.Monad
import Data.Bifunctor
import Data.List as L
import Data.Set as S
import Data.Maybe

solve = traverse (uncurry (flip isPrefixOf')) . pairs

pairs = uncurry fmap . first (,)
    <=< mapMaybe uncons . tails . fmap snd . S.toDescList . S.fromList . fmap (\s -> (length s, s))

isPrefixOf' xs ys | xs `isPrefixOf` ys = Left ys
                  | otherwise          = Right ()

parse = L.drop 1 . lines

run = solve . parse

main = interact $ either ("BAD SET\n" <>) (const "GOOD SET") . run
